src/cache.c

```diff
@@ -0,0 +1,378 @@
+/*
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
+#include "cache.h"
+#include <assert.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define MIN(a, b) ((a < b) ? a : b)
+#define GOLDEN_RATIO_32 0x61C88647
+#define HASH(val) \
+    (((val) * (GOLDEN_RATIO_32)) >> (32 - (cache_size_bits))) % (cache_size)
+
+static uint32_t cache_size, cache_size_bits;
+
+typedef enum {
+    LRU_list,
+    LFU_list,
+    LRU_ghost_list,
+    LFU_ghost_list,
+    list_quantity
+} list_category_t;
+
+struct list_head {
+    struct list_head *prev, *next;
+};
+
+struct hlist_head {
+    struct hlist_node *first;
+};
+
+struct hlist_node {
+    struct hlist_node *next, **pprev;
+};
+
+typedef struct arc_entry {
+    void *value;
+    uint32_t key;
+    list_category_t arc_type;
+    struct list_head list;
+    struct hlist_node ht_list;
+} arc_entry_t;
+
+typedef struct hashtable {
+    struct hlist_head *ht_list_head;
+} hashtable_t;
+
+typedef struct cache {
+    struct list_head *list_table[list_quantity];
+    uint32_t list_size[list_quantity];
```

Comment:
What are typical ranges for the lists used in ARC? I don't think it is worthy to preserve them in `uint32_t`.

Reply:
The maximum of list size T1, T2, and B1 is equal to `cache_size / 2`, and the maximum of list size B2 is equal to `cache_size`. Therefore, if user set `cache_size` to more than 2^17, we would need uint32_t to store list_size. Otherwise, it might cause overflow.